File: app/src/main/java/com/benny/openlauncher/widget/Dock.java
package com.benny.openlauncher.widget;

import android.content.Context;
import android.graphics.Point;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.benny.openlauncher.activity.HomeActivity;
import com.benny.openlauncher.manager.Setup;
import com.benny.openlauncher.model.Item;
import com.benny.openlauncher.util.Definitions.ItemPosition;
import com.benny.openlauncher.util.DragAction.Action;
import com.benny.openlauncher.util.DragHandler;
import com.benny.openlauncher.util.Tool;
import com.benny.openlauncher.viewutil.DesktopCallback;
import com.benny.openlauncher.viewutil.ItemViewFactory;

import java.util.List;

public final class Dock extends CellContainer implements DesktopCallback {
    private HomeActivity _homeActivity;
    private final Point _coordinate = new Point();
    private final Point _previousDragPoint = new Point();

    private Item _previousItem;
    private View _previousItemView;

    // open app drawer on slide up gesture
    private float _startPosX;
    private float _startPosY;

    public Dock(Context context, AttributeSet attr) {
        super(context, attr);
    }

    public final void initDock() {
        int columns = Setup.appSettings().getDockColumnCount();
        int rows = Setup.appSettings().getDockRowCount();
        setGridSize(columns, rows);
        List<Item> dockItems = HomeActivity._db.getDock();
        removeAllViews();
        for (Item item : dockItems) {
            if (item._x + item._spanX <= columns && item._y + item._spanY <= rows) {
                addItemToPage(item, 0);
            }
        }

        // call onMeasure to set the height
        measure(getMeasuredWidth(), getMeasuredHeight());
    }

    public boolean dispatchTouchEvent(@NonNull MotionEvent ev) {
        detectSwipe(ev);
        super.dispatchTouchEvent(ev);
        return true;
    }

    private void detectSwipe(MotionEvent ev) {
        switch (ev.getAction()) {
            case 0:
                _startPosX = ev.getX();
                _startPosY = ev.getY();
                break;
            case 1:
                if (_startPosY - ev.getY() > 150.0f && Setup.appSettings().getGestureDockSwipeUp()) {
                    Point point = new Point((int) ev.getX(), (int) ev.getY());
                    point = Tool.convertPoint(point, this, _homeActivity.getAppDrawerController());
                    if (Setup.appSettings().getGestureFeedback()) {
                        Tool.vibrate(this);
                    }
                    _homeActivity.openAppDrawer(this, point.x, point.y);
                    break;
                }
            default:
                break;
        }
    }

    public final void updateIconProjection(int x, int y) {
        HomeActivity launcher = _homeActivity;
        ItemOptionView dragNDropView = launcher.getItemOptionView();
        DragState state = peekItemAndSwap(x, y, _coordinate);
        if (!_coordinate.equals(_previousDragPoint)) {
            dragNDropView.cancelFolderPreview();
        }
        _previousDragPoint.set(_coordinate.x, _coordinate.y);
        switch (state) {
            case CurrentNotOccupied:
                projectImageOutlineAt(_coordinate, DragHandler._cachedDragBitmap);
                break;
            case CurrentOccupied:
                Item.Type type = dragNDropView.getDragItem()._type;
                clearCachedOutlineBitmap();
                if (!type.equals(Item.Type.WIDGET) && (coordinateToChildView(_coordinate) instanceof AppItemView)) {
                    dragNDropView.showFolderPreviewAt(this, getCellWidth() * (_coordinate.x + 0.5f), getCellHeight() * (_coordinate.y + 0.5f) - (Setup.appSettings().getDockShowLabel() ? Tool.dp2px(7) : 0));
                }
                break;
            case OutOffRange:
            case ItemViewNotFound:
            default:
                break;
        }
    }

    @Override
    public void setLastItem(Item item, View view) {
        _previousItemView = view;
        _previousItem = item;
        removeView(view);
    }

    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (!isInEditMode()) {
            // set the height for the dock based on the number of rows and the show label preference
            int iconSize = Setup.appSettings().getDockIconSize();
            int height = Tool.dp2px((iconSize + 20) * getCellSpanV());
            if (Setup.appSettings().getDockShowLabel()) height += Tool.dp2px(20);
            getLayoutParams().height = height;
            setMeasuredDimension(View.getDefaultSize(getSuggestedMinimumWidth(), widthMeasureSpec), height);
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        }
    }

    @Override
    public void consumeLastItem() {
        _previousItem = null;
        _previousItemView = null;
    }

    @Override
    public void revertLastItem() {
        if (_previousItemView != null && _previousItem != null) {
            addViewToGrid(_previousItemView);
            _previousItem = null;
            _previousItemView = null;
        }
    }

    public boolean addItemToPage(@NonNull Item item, int page) {
        View itemView = ItemViewFactory.getItemView(getContext(), this, Action.DESKTOP, item, isDockShowLabel());
        if (itemView == null) {
            // TODO see if this fixes SD card bug
            //HomeActivity._db.deleteItem(item, true);
            return false;
        }
        item._location = ItemPosition.Dock;
        addViewToGrid(itemView, item._x, item._y, item._spanX, item._spanY);
        return true;
    }

    public boolean addItemToPoint(@NonNull Item item, int x, int y) {
        LayoutParams positionToLayoutPrams = coordinateToLayoutParams(x, y, item._spanX, item._spanY);
        if (positionToLayoutPrams == null) {
            return false;
        }
        item._location = ItemPosition.Dock;
        item._x = positionToLayoutPrams.getX();
        item._y = positionToLayoutPrams.getY();
        View itemView = ItemViewFactory.getItemView(getContext(), this, Action.DESKTOP, item, isDockShowLabel());
        if (itemView != null) {
            itemView.setLayoutParams(positionToLayoutPrams);
            addView(itemView);
        }
        return true;
    }

    public boolean addItemToCell(@NonNull Item item, int x, int y) {
        item._location = ItemPosition.Dock;
        item._x = x;
        item._y = y;
        View itemView = ItemViewFactory.getItemView(getContext(), this, Action.DESKTOP, item, isDockShowLabel());
        if (itemView == null) {
            return false;
        }
        addViewToGrid(itemView, item._x, item._y, item._spanX, item._spanY);
        return true;
    }

    public void removeItem(final View view, boolean animate) {
        if (animate) {
            view.animate().setDuration(100).scaleX(0.0f).scaleY(0.0f).withEndAction(new Runnable() {
                @Override
                public void run() {
                    if (view.getParent().equals(Dock.this)) {
                        removeView(view);
                    }
                }
            });
        } else if (this.equals(view.getParent())) {
            removeView(view);
        }
    }

    public void setHome(HomeActivity homeActivity) {
        _homeActivity = homeActivity;
    }

    private Boolean isDockShowLabel() {
        boolean b = Setup.appSettings().getDockShowLabel();
        Boolean ret = new Boolean(b);
        return ret;
    }
}

File: app/src/main/java/com/benny/openlauncher/widget/Dock.kt
package com.benny.openlauncher.widget

import android.content.Context
import android.graphics.Point
import android.os.Build
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.WindowInsets
import com.benny.openlauncher.activity.CoreHome
import com.benny.openlauncher.manager.Setup
import com.benny.openlauncher.model.Item
import com.benny.openlauncher.util.DragAction
import com.benny.openlauncher.util.DragNDropHandler
import com.benny.openlauncher.util.Tool
import com.benny.openlauncher.viewutil.DesktopCallBack
import com.benny.openlauncher.viewutil.ItemViewFactory

class Dock @JvmOverloads constructor(c: Context, attr: AttributeSet? = null) : CellContainer(c, attr), DesktopCallBack<View> {
    var previousItemView: View? = null
    var previousItem: Item? = null
    private var startPosX: Float = 0.toFloat()
    private var startPosY: Float = 0.toFloat()
    private var home: CoreHome? = null
    private val coordinate = Point()

    override fun init() {
        if (isInEditMode) {
            return
        }
        super.init()
    }

    fun initDockItem(home: CoreHome) {
        val columns = Setup.appSettings().dockSize
        setGridSize(columns, 1)
        val dockItems = CoreHome.db.dock

        this.home = home
        removeAllViews()
        for (item in dockItems) {
            if (item.x < columns && item.y == 0) {
                addItemToPage(item, 0)
            }
        }
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean = super.onTouchEvent(ev)

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        detectSwipe(ev)
        super.dispatchTouchEvent(ev)
        return true
    }

    private fun detectSwipe(ev: MotionEvent) {
        if (CoreHome.launcher == null) return
        when (ev.action) {
            MotionEvent.ACTION_UP -> {
                Tool.print("ACTION_UP")
                val minDist = 150f
                Tool.print(ev.x.toInt(), ev.y.toInt())
                if (startPosY - ev.y > minDist) {
                    if (Setup.appSettings().gestureDockSwipeUp) {
                        val p = Tool.convertPoint(Point(ev.x.toInt(), ev.y.toInt()), this, CoreHome.launcher!!.getAppDrawerController())
                        if (Setup.appSettings().isGestureFeedback)
                            Tool.vibrate(this)
                        CoreHome.launcher!!.openAppDrawer(this, p.x, p.y)
                    }
                }
            }
            MotionEvent.ACTION_DOWN -> {
                Tool.print("ACTION_DOWN")
                startPosX = ev.x
                startPosY = ev.y
            }
        }
    }

    private val previousDragPoint = Point()

    fun updateIconProjection(x: Int, y: Int) {
        val state = peekItemAndSwap(x, y, coordinate)

        if (previousDragPoint != coordinate)
            CoreHome.launcher?.getDragNDropView()?.cancelFolderPreview()
        previousDragPoint.set(coordinate.x, coordinate.y)

        when (state) {
            CellContainer.DragState.CurrentNotOccupied -> projectImageOutlineAt(coordinate, DragNDropHandler.cachedDragBitmap)
            CellContainer.DragState.OutOffRange -> {

            }
            CellContainer.DragState.ItemViewNotFound -> {
            }
            CellContainer.DragState.CurrentOccupied -> {
                clearCachedOutlineBitmap()

                val action = CoreHome.launcher?.getDragNDropView()?.dragAction
                if (action != DragAction.Action.WIDGET && action != DragAction.Action.ACTION &&
                        coordinateToChildView(coordinate) is AppItemView)
                    CoreHome.launcher?.getDragNDropView()?.showFolderPreviewAt(
                            this,
                            cellWidth * (coordinate.x + 0.5f),
                            cellHeight * (coordinate.y + 0.5f) - if (Setup.appSettings().isDockShowLabel) Tool.toPx(7) else 0
                    )
            }
        }
    }

    override fun setLastItem(vararg args: Any) {
        // args stores the item in [0] and the view reference in [1]
        val v = args[1] as View
        val item = args[0] as Item

        previousItemView = v
        previousItem = item
        removeView(v)
    }

    override fun onApplyWindowInsets(insets: WindowInsets): WindowInsets {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            bottomInset = insets.systemWindowInsetBottom
            setPadding(paddingLeft, paddingTop, paddingRight, bottomInset)
        }
        return insets
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (isInEditMode) return

        var height = View.getDefaultSize(suggestedMinimumHeight, heightMeasureSpec)

        val iconSize = Setup.appSettings().dockIconSize
        if (Setup.appSettings().isDockShowLabel) {
            height = Tool.dp2px(16 + iconSize + 14 + 10, context) + Dock.bottomInset
        } else {
            height = Tool.dp2px(16 + iconSize + 10, context) + Dock.bottomInset
        }
        layoutParams.height = height

        setMeasuredDimension(View.getDefaultSize(suggestedMinimumWidth, widthMeasureSpec), height)

        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    override fun consumeRevert() {
        previousItem = null
        previousItemView = null
    }

    override fun revertLastItem() {
        if (previousItemView != null) {
            addViewToGrid(previousItemView!!)
            previousItem = null
            previousItemView = null
        }
    }

    override fun addItemToPage(item: Item, page: Int): Boolean {
        val itemView = ItemViewFactory.getItemView(context, item, Setup.appSettings().isDockShowLabel, this, Setup.appSettings().dockIconSize)

        if (itemView == null) {
            CoreHome.db.deleteItem(item, true)
            return false
        } else {
            item.locationInLauncher = Item.LOCATION_DOCK
            addViewToGrid(itemView, item.x, item.y, item.spanX, item.spanY)
            return true
        }
    }

    override fun addItemToPoint(item: Item, x: Int, y: Int): Boolean {
        val positionToLayoutPrams = coordinateToLayoutParams(x, y, item.spanX, item.spanY)
        if (positionToLayoutPrams != null) {
            item.locationInLauncher = Item.LOCATION_DOCK

            item.x = positionToLayoutPrams.x
            item.y = positionToLayoutPrams.y

            val itemView = ItemViewFactory.getItemView(context, item, Setup.appSettings().isDockShowLabel, this, Setup.appSettings().dockIconSize)

            if (itemView != null) {
                itemView.layoutParams = positionToLayoutPrams
                addView(itemView)
            }
            return true
        } else {
            return false
        }
    }

    override fun addItemToCell(item: Item, x: Int, y: Int): Boolean {
        item.locationInLauncher = Item.LOCATION_DOCK

        item.x = x
        item.y = y

        val itemView = ItemViewFactory.getItemView(context, item, Setup.appSettings().isDockShowLabel, this, Setup.appSettings().dockIconSize)

        if (itemView != null) {
            addViewToGrid(itemView, item.x, item.y, item.spanX, item.spanY)
            return true
        } else {
            return false
        }
    }

    override fun removeItem(view: View, animate: Boolean) {
        if (animate)
            view.animate().setDuration(100L).scaleX(0f).scaleY(0f).withEndAction({
                if (view.parent == this)
                    removeView(view)
            })
        else
            if (view.parent == this)
                removeView(view)
    }

    companion object {
        var bottomInset: Int = 0
    }
}
